### ISSUE TITLE
Separate compilation not working on Windows

### ISSUE THREAD

**Issue by kmaehashi** [cupy/cupy#4421]
The following code runs fine on Linux but fails on Windows:

```py
import cupy
kernel = cupy.RawKernel('extern "C" __global__ void test() {}', 'test', backend='nvcc', enable_cooperative_groups=True)
kernel.compile()
```

Error raised in Windows:

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "cupy\core\raw.pyx", line 282, in cupy.core.raw.RawKernel.compile
  File "cupy\core\raw.pyx", line 110, in cupy.core.raw.RawKernel._kernel
  File "cupy\cuda\function.pyx", line 234, in cupy.cuda.function.Module.get_function
  File "cupy\cuda\function.pyx", line 175, in cupy.cuda.function.Function.__init__
  File "cupy_backends\cuda\api\driver.pyx", line 262, in cupy_backends.cuda.api.driver.moduleGetFunction
  File "cupy_backends\cuda\api\driver.pyx", line 124, in cupy_backends.cuda.api.driver.check_status
cupy_backends.cuda.api.driver.CUDADriverError: CUDA_ERROR_NOT_FOUND: named symbol not found
```

Internally, the above code does the following nvcc separate compilation:

```
nvcc -gencode=arch=compute_70,code=sm_70 -ftz=true -rdc=true -Xcompiler -fPIC -shared -o test_rdc.o test.cu
nvcc -gencode=arch=compute_70,code=sm_70 --cubin -ftz=true -rdc=true -Xcompiler -fPIC -shared test_rdc.o --device-link -o test_rdc.cubin
```

It generates a correct cubin on Linux:

```
% cuobjdump -symbols test_rdc.cubin

symbols:
STT_FUNC         STB_GLOBAL STO_ENTRY      test
```

But on Windows it generates an empty (?) cubin, which is causing `named symbol not found` error:

```
C:\Users\chainer\Documents>cuobjdump -all -symbols test_rdc.cubin

symbols:
```

It seems we need some other commands to do separate compilation on Windows.

cc/ @leofang @anaruse

**Comment by anaruse** [comment]
This might be a dumb question, but why do we need a separate compilation?

**Comment by leofang** [comment]
Two examples are cooperative group or dynamic parallelism, where a device linking (to the static device runtime library) must happen, and so a separate compilation is required.

**Comment by leofang** [comment]
@kmaehashi I found CMake provides an option `CUDA_RESOLVE_DEVICE_SYMBOLS` to address this, but I don't know which compiler flag(s) it translates to: https://stackoverflow.com/a/51353324/2344149

**Comment by anaruse** [comment]
I'm not familiar with separate compilation, but what happens if you compile with the following options?

    nvcc -gencode=arch=compute_70,code=sm_70 -v --device-c -ftz=true test.cu -o test.o 
    nvcc -gencode=arch=compute_70,code=sm_70 -v --device-link --cubin -ftz=true test.o -o test.cubin

**Comment by leofang** [comment]
I think it's what we do?

**Comment by anaruse** [comment]
I don't know well about Windows, but the Windows compiler probably dosen't have a -fPIC option, does it?

    nvcc -gencode=arch=compute_70,code=sm_70 -ftz=true -rdc=true -Xcompiler -fPIC -shared -o test_rdc.o test.cu
    nvcc -gencode=arch=compute_70,code=sm_70 --cubin -ftz=true -rdc=true -Xcompiler -fPIC -shared test_rdc.o --device-link -o test_rdc.cubin

**Comment by leofang** [comment]
Interesting...Yes for compiling the CUDA codes at build time, I do notice that MSVC does not use `-fPIC` to `postargs`:
https://github.com/cupy/cupy/blob/fa3a87494068d46816f7305735cfa3661cc568b3/cupy_setup_build.py#L1014
So perhaps we could give it a shot? Though if it's the problem, nvcc should have raised an error...

**Comment by kmaehashi** [comment]
`--device-c` magically solved the issue! (#4470) Thank you very much 👍🏼

**Comment by kmaehashi** [comment]
It seems that `--device-c` option support in ccache 3.7.12 is incomplete.  I haven't tested other ccache versions, but sharing FYI for developers.

```
% touch test.cu
% nvcc --ptx --device-c test.cu
% ll
total 4
-rw-rw-r-- 1 maehashi maehashi   0 Dec 22 11:18 test.cu
-rw-rw-r-- 1 maehashi maehashi 203 Dec 22 11:18 test.ptx
% ccache nvcc --ptx --device-c test.cu
% ll
total 8
-rw-rw-r-- 1 maehashi maehashi   0 Dec 22 11:18 test.cu
-rw-rw-r-- 1 maehashi maehashi 203 Dec 22 11:18 test.o   # generates test.o instead of test.ptx
-rw-rw-r-- 1 maehashi maehashi 203 Dec 22 11:18 test.ptx
```
